### DISCUSSION TITLE
[Server actions] Error: Body exceeded 1mb limit

### DISCUSSION THREAD

**Discussion by kasipavankumar** [vercel/next.js#49891] [Help] [answered]
### Summary

I am using server actions for file uploads. When trying to upload a file bigger than 1MB, the following error is being raised by the Next server:

```bash
error Error: Body exceeded 1mb limit
```

I tried the old way of exporting config, which seem to be deprecated.

```js
export const config = {
    api: {
        bodyParser: {
            sizeLimit: '8mb',
        },
    },
}
```

Is there a way I can increase the body size limit for server actions?

**Comment by happypoo35** [comment]
As of now I don't see any way to set body limit programmatically.

For testing purposes you can change limit value in `node_modules\next\dist\server\app-render\action-handler.js`
```js
const actionData = await parseBody(req, "10mb") || ""; // Changed limit to 10mb
```

**Reply by praveenswadesh** [reply]
As per [this line](https://github.com/vercel/next.js/blob/5526ea504c01135dbcdee2c0c2a26de43252c108/packages/next/src/server/api-utils/node.ts#L556), Shouldn't this be
  
```
  config.api.bodyParser.sizeLimit
```

**Reply by itsjavi** [reply]
this has no effect. it seems hardcoded to '1mb' https://github.com/vercel/next.js/blob/canary/packages/next/src/server/app-render/action-handler.ts#L331

I hope we get a configurable option in next version

**Comment by devjiwonchoi** [comment]
This feature has been blocked since `App Router`. It is currently hard-coded to `1mb` as you can see [here](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/app-render/action-handler.ts#L355).

You need to customize the `parseBody()` size manually through `node_modules`.. not recommended but no "quicker" option is available at this moment.

The file is located in `node_modules\next\dist\server\app-render\action-handler.js`.

![Screenshot 2023-06-11 at 12 55 49 AM](https://github.com/vercel/next.js/assets/120007119/0447377a-31b3-4934-9856-f59aeca36c58)

I've opened PR #51104 to resolve this hard-coded size issue.

## UPDATE

Resolved with #51104

## Reference
- [API Routes Response Size Limit](https://nextjs.org/docs/messages/api-routes-response-size-limit)
- [Action Handler Set to `1mb`](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/app-render/action-handler.ts#L355)

**Reply by itsjavi** [reply]
thank you! I was going to also open a PR but got no time :(

**Reply by devjiwonchoi** [reply, marked as answer]
@kasipavankumar `serverActiionsBodySizeLimit` has been added on [release v13.4.8](https://github.com/vercel/next.js/releases/tag/v13.4.8).

**Comment by PatrickTimas** [comment]
on your next.config.js 

```
module.exports = {
  experimental: {
    serverActions: {
      allowedOrigins: ['my-proxy.com', '*.my-proxy.com'],
    },
  },
}
```

From: https://nextjs.org/docs/app/api-reference/next-config-js/serverActions

**Comment by Marabii** [comment]
there's always a maximum amount of data that can uploaded using server actions, and if surpassed it throws an error to the console and there's no way to catch it and handle it gracefully. Correct me if i'm wrong.  I tried this: 

  const handleSubmit = async (formData: FormData) => {
    try {
      formAction(formData);
    } catch (e) {
      alert("something went wrong");
    }
  };
  
  i then pass handleSubmit as an action to the form but it doesn't work.

**Reply by geekysaurabh001** [reply]
As @Marabii said. No way to catch it. This gives Application Error in Production.

**Comment by roofleDev** [comment]
I spent an hour trying to catch this error but regardless of including the try/catch blocks client-side and server-side, I can't catch it. When my client component calls the server action it throws the uncaught error on the server and hangs until the timeout limit.

The server logs:
```
 uncaughtException:  [Error: Body exceeded 1 MB limit.
                To configure the body size limit for Server Actions, see: https://nextjs.org/docs/app/api-reference/next-config-js/serverActions#bodysizelimit] {
  statusCode: 413
}
```

I have increased my server action body size limit, but it is unfortunate this error can't seem to be caught and reported back to the frontend in a meaningful way. If I'm wrong and there is a way, please let us know.

**Reply by jcamatta** [reply]
handleUploadFile = (e) => { file = e.target.files[0] if (file.size > MAX_LIMIT) { alert } }

**Reply by roofleDev** [reply]
Thanks @jcamatta, sometimes I can't see the forest through the trees. That's a nice solution.
